### PR TITLE
Backtracking support through &try/backtrack()

### PR DESCRIPTION
**Pull request by rsmmr**

### CODE REVIEW

**Review by bbannier — COMMENTED** [review]
This looks like a great tool so users can build their own tooling to parse conditionally. At the same time, I could see this being used to implement error handling for which this patch doesn't seem to go all the way or at least is missing tests (e.g., can `backtrack` be caught by `on %error` handlers; guaranteed interaction of plain old parse errors and `&try` fields; no access to particular error when `&try` field fails to parse). I left a few comments along those lines below. Like discussed offline we probably still need a better idea how we want error handling to finally look like.

File: tests/spicy/types/unit/backtrack.spicy [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    foo: Foo &try;
+    bar: Bar;
+    b: bytes &size=6;
+
+    on %done { print self; }
+};
+
+type Foo = unit {
+    a: int8 { print "Foo.a", self; }
+    b: int8 { print "Backtracking"; self.backtrack(); }
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply tests/spicy/types/unit/backtrack.spicy:17]: The print in the final `%done` is actually already showing the partial construction.

File: doc/programming/parsing.rst [hunk 2]
```diff
@@ -1213,6 +1216,56 @@ once you have subunits that are recognizable by how they start:
...
+
+Backtracking
+^^^^^^^^^^^^
+
+Spicy supports a simple form of manual backtracking. If a field is
+marked with ``&try``, a later call to the unit's ``backtrack()``
+method anywhere down in the parse tree originating at that field, will
+immediately transfer control over to the field following the ``&try``.
+When doing so, the data position inside the input stream will be reset to
+where it was when the ``&try`` field started its processing. Example:
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply doc/programming/parsing.rst:1229]: Not sure what to add to the text to make it more clear, can you suggest some wording?

File: spicy/toolchain/include/ast/operators/unit.h [hunk 3]
```diff
@@ -116,4 +116,18 @@ this method will not do anything.
...
 
+BEGIN_METHOD(unit, Backtrack)
+    auto signature() const {
+        return hilti::operator_::Signature{.self = hilti::type::constant(spicy::type::Unit(type::Wildcard())),
+                                           .result = hilti::type::Void(),
+                                           .id = "backtrack",
+                                           .args = {},
+                                           .doc = R"(
+Aborts parsing at the current position and returns back to the most recent
+``&try`` attribute. Turns into a parse error if there's no ``&try`` in scope.
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply spicy/toolchain/include/ast/operators/unit.h:127]: Yeah, that's how it's meant to be used, though I wouldn't frame it directly as triggering a parse error (that's just what happens internally), but as "initiating backtracking". We can (and should) add an explicit `raise` of some kind as well, I'm just not quite sure on syntax yet, suggestions welcome (would probably also take a string to explain what's going on).

File: tests/spicy/types/unit/backtrack-lah.spicy [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+type Line = unit {
+    switch {
+        -> a: A;
+        -> b: B;
+        -> c: C;
+    };
+
+    on %error {
+        self.backtrack();
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply tests/spicy/types/unit/backtrack-lah.spicy:29]: It's intentional to make this explicit/manual here for this backtracking. We can add further functionality for more automated recovery.

File: doc/programming/parsing.rst (hunk 2 above)
[the remark this answers is not in the document]
Comment (bbannier) [reply doc/programming/parsing.rst:1229]: Probably hinting that each `&try` adds a jump point to a stack and that that stack is unwound using `backtrack` would already help, e.g.,

> ... transfer control over to the field following the ``&try``; if incompletely parsed fields with `&try` are nested when `backtrack()` is called we jump back to the innermost unparsed field. When doing so ...

Something which explains nesting semantics without becoming too technical.

**Review by bbannier — DISMISSED** [review]

**Pull request merged**